File: Walker App/KAppMakerExtended-main/iosApp/iosApp/Ads/Rewarded/RewardedAdDisplayer.swift
//
//  RewardedAdDisplayer.swift
//  iosApp
//
//  Created by Mirzamehdi on 07/03/2025.
//

import Foundation
import ComposeApp
import SwiftUI
import GoogleMobileAds


class RewardedAdDisplayer: NSObject, FullScreenAdDisplayer {
    
    private let adLoader: FullScreenAdLoader
    private let onRewarded: (AdsRewardItem) -> Void
    
    init(adLoader: FullScreenAdLoader, onRewarded: @escaping (AdsRewardItem) -> Void) {
        self.adLoader = adLoader
        self.onRewarded = onRewarded
    }
    
    func show() {
        
        guard let rewardedAdLoader = adLoader as? RewardedAdLoader else {
            print("Ad loader is not of type RewardedAdLoader")
            return
        }
        
        guard let rewardedAd = rewardedAdLoader.rewardedAd else {
            print("Rewarded ad is not loaded yet")
            rewardedAdLoader.load()
            return
        }
        
        rewardedAd.fullScreenContentDelegate = self
        rewardedAd.present(fromRootViewController: nil) {
            let reward = rewardedAd.adReward
            let rewardItem = AdsRewardItem(
                amount: reward.amount.int32Value,
                type: reward.type
            )
            self.onRewarded(rewardItem)
            print("Reward received with type \(reward.type), amount \(reward.amount)")
          }
        
    }
    
}

extension RewardedAdDisplayer: GADFullScreenContentDelegate {
    
    func adDidDismissFullScreenContent(_ ad: any GADFullScreenPresentingAd) {
        print("Rewarded ad is dismissed, loading a new one")
        guard let rewardedAdLoader = adLoader as? RewardedAdLoader else { return }
        rewardedAdLoader.rewardedAd = nil
        rewardedAdLoader.load()
    }
    
    func ad(_ ad: any GADFullScreenPresentingAd, didFailToPresentFullScreenContentWithError error: any Error) {
        print("Rewarded ad failed to show")
        guard let rewardedAdLoader = adLoader as? RewardedAdLoader else { return }
        rewardedAdLoader.rewardedAd = nil
    }
    
    func adWillPresentFullScreenContent(_ ad: any GADFullScreenPresentingAd) {
        print("Rewarded ad is shown")
    }
    
}

File: Walker App/KAppMakerExtended-main/composeApp/src/androidMain/kotlin/com/measify/kappmaker/presentation/components/ads/rewarded/RewardedAdDisplayer.kt
package com.measify.kappmaker.presentation.components.ads.rewarded

import androidx.activity.ComponentActivity
import com.google.android.gms.ads.AdError
import com.google.android.gms.ads.FullScreenContentCallback
import com.google.android.gms.ads.OnUserEarnedRewardListener
import com.measify.kappmaker.presentation.components.ads.AdsRewardItem
import com.measify.kappmaker.presentation.components.ads.FullScreenAdDisplayer
import com.measify.kappmaker.presentation.components.ads.FullScreenAdLoader
import com.measify.kappmaker.util.logging.AppLogger

class RewardedAdDisplayer(
    private val activity: ComponentActivity?,
    private val adLoader: FullScreenAdLoader,
    private val onRewarded: (AdsRewardItem) -> Unit,
) : FullScreenAdDisplayer {

    override fun show() {
        if (adLoader !is RewardedAdLoader) return
        val rewardedAd = adLoader.rewardedAd

        if (rewardedAd == null || activity == null) {
            AppLogger.d("Rewarded ad is not loaded yet")
            adLoader.load()
            return
        }

        rewardedAd.fullScreenContentCallback = object : FullScreenContentCallback() {
            override fun onAdDismissedFullScreenContent() {
                super.onAdDismissedFullScreenContent()
                AppLogger.d("Rewarded ad is dismissed, loading new one")
                adLoader.rewardedAd = null
                adLoader.load()
            }

            override fun onAdFailedToShowFullScreenContent(p0: AdError) {
                super.onAdFailedToShowFullScreenContent(p0)
                AppLogger.e("Rewarded ad, failed to show: ${p0.message}")
                adLoader.rewardedAd = null
            }

            override fun onAdShowedFullScreenContent() {
                super.onAdShowedFullScreenContent()
                AppLogger.d("Rewarded ad is shown")
            }
        }
        rewardedAd.show(activity, OnUserEarnedRewardListener { rewardItem ->
            val adsRewardItem = AdsRewardItem(
                amount = rewardItem.amount,
                type = rewardItem.type
            )

            onRewarded(adsRewardItem)
            AppLogger.d("User earned reward: amount: ${rewardItem.amount}, type: ${rewardItem.type}")

        })

    }
}